### PR TITLE
fix(utils/atomWithStorage): defaultStorage with disabled local storage crashes on mount

### PR DESCRIPTION
**Pull request by benediktschlager**
## Related Issues or Discussions

Fixes downstream issue for https://ui.shadcn.com/
It crashes when cookies are disabled with this exception:
image

## Summary
On Firefox accessing the property localStorage of window results in a `DOMException: The operation is insecure.`
This PR will handle this case gracefully and fall back to providing no storage implementation. This should be fine as we cannot store anything in a cookie-free browser.

## Check List

- [x] `yarn run prettier` for formatting code and docs

### CODE REVIEW

File: src/vanilla/utils/atomWithStorage.ts [hunk 1]
```diff
@@ -99,8 +99,18 @@ export function createJSONStorage<Value>(
...
     | AsyncStringStorage
     | SyncStringStorage
-    | undefined = () =>
-    typeof window !== 'undefined' ? window.localStorage : undefined,
+    | undefined = () => {
+    try {
+      const storage =
+        typeof window !== 'undefined' ? window.localStorage : undefined
+      return storage
+    } catch (e: DOMException | any) {
```
Comment (dai-shi) [inline src/vanilla/utils/atomWithStorage.ts:107]:
```suggestion
    } catch (e) {
```
should work?

File: src/vanilla/utils/atomWithStorage.ts [hunk 2]
```diff
@@ -99,8 +99,18 @@ export function createJSONStorage<Value>(
   getStringStorage: () =>
     | AsyncStringStorage
     | SyncStringStorage
-    | undefined = () =>
-    typeof window !== 'undefined' ? window.localStorage : undefined,
+    | undefined = () => {
+    try {
+      const storage =
+        typeof window !== 'undefined' ? window.localStorage : undefined
+      return storage
```
Comment (dai-shi) [inline src/vanilla/utils/atomWithStorage.ts:106]: Can we do this and catch "undefined" error below?
```suggestion
      return window.localStorage
```

File: tests/react/vanilla-utils/atomWithStorage.test.tsx [hunk 3]
```diff
@@ -502,6 +502,44 @@ describe('atomWithStorage (with browser storage)', () => {
   })
 })
 
+describe('atomWithStorage (with disabled browser storage)', () => {
+  const localStorage = window.localStorage
```
Comment (dai-shi) [inline tests/react/vanilla-utils/atomWithStorage.test.tsx:506]: Just a preference.
```suggestion
  const savedLocalStorage = window.localStorage
```

**Review by dai-shi — CHANGES_REQUESTED** [review]
Hi! This looks like a nice improvement!
Please check comments.

**Comment by dai-shi** [comment]
> On Firefox accessing the property localStorage of window results in a `DOMException: The operation is insecure.`

On Chrome, it's the same with slightly different message:
`DOMException: Failed to read the 'localStorage' property from 'Window': Access is denied for this document.`

**Comment by dai-shi** [comment]
@benediktschlager hi, are you still around?

File: src/vanilla/utils/atomWithStorage.ts [hunk 4]
```diff
@@ -99,8 +99,18 @@ export function createJSONStorage<Value>(
...
     | AsyncStringStorage
     | SyncStringStorage
-    | undefined = () =>
-    typeof window !== 'undefined' ? window.localStorage : undefined,
+    | undefined = () => {
+    try {
+      return window.localStorage
+    } catch (e) {
+      // Window could be undefined or local storage could be disabled
+      if (e instanceof TypeError || e instanceof DOMException) {
```
Comment (dai-shi) [inline src/vanilla/utils/atomWithStorage.ts:107]: Hmm, this is what I tried in Node.js:
```
> try { window.localStorage } catch (e) { console.log(e instanceof TypeError) }
false
undefined
> try { window.localStorage } catch (e) { console.log(e instanceof ReferenceError) }
true
undefined
```
Maybe, it's not stable. On the other hand, when is the case we want to `throw e`?

**Comment by benediktschlager** [comment]
Hi @dai-shi, I included your feedback.
Can you check again?
Some non-required steps in the pipeline are failing.
Is this fine, or should I add a _as any_ cast to the test to mute this error?

File: src/vanilla/utils/atomWithStorage.ts [hunk 5]
```diff
@@ -99,8 +99,18 @@ export function createJSONStorage<Value>(
...
+        return undefined
+      }
+
+      throw e
+    }
```
Comment (dai-shi) [inline src/vanilla/utils/atomWithStorage.ts:112]: completely different idea:
```suggestion
    try {
      return window.localStorage
    } catch (e) {
      if (import.meta.env?.MODE !== 'production') {
        if (typeof window !== 'undefined') {
          console.warn(e);
        }
      }
      return undefined
    }
```

File: src/vanilla/utils/atomWithStorage.ts (hunk 4 above)
Comment (benediktschlager) [reply src/vanilla/utils/atomWithStorage.ts:107]: Oh, it seems on node the window instance is not even defined.
I'm wondering under which circumstances the window instance is not defined, but this library is used?
As far as I understand, Jotai only works alongside React?

File: tests/react/vanilla-utils/atomWithStorage.test.tsx [hunk 6]
```diff
@@ -502,6 +502,44 @@ describe('atomWithStorage (with browser storage)', () => {
...
+    // Firefox and chromium based browser throw DOMException when cookies are disabled
+    Object.defineProperty(window, 'localStorage', {
+      get() {
+        throw new DOMException('The operation is insecure.')
+      },
+    })
+  })
+
+  afterAll(() => {
+    window.localStorage = savedLocalStorage
```
Comment (dai-shi) [inline tests/react/vanilla-utils/atomWithStorage.test.tsx:518]:
```suggestion
    // TS < 4.5 causes type error
    (window as any).localStorage = savedLocalStorage
```

File: src/vanilla/utils/atomWithStorage.ts (hunk 4 above)
Comment (dai-shi) [reply src/vanilla/utils/atomWithStorage.ts]: SSR=Server Side Rendering.

**Comment by dai-shi** [comment]
oops, can you run prettier please?

File: tests/react/vanilla-utils/atomWithStorage.test.tsx [hunk 7]
```diff
@@ -502,6 +502,45 @@ describe('atomWithStorage (with browser storage)', () => {
...
+    // Firefox and chromium based browser throw DOMException when cookies are disabled
+    Object.defineProperty(window, 'localStorage', {
+      get() {
+        throw new DOMException('The operation is insecure.')
+      },
+    })
+  })
+
+  afterAll(() => {
+    // TS < 4.5 causes type error
```
Comment (dai-shi) [inline tests/react/vanilla-utils/atomWithStorage.test.tsx:518]:
```suggestion
    // TS < 4.5 causes type error without `as any`
```

**Review by dai-shi — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**